wasm-wrappers/wasm-doc-gen/src/main.rs

```diff
@@ -0,0 +1,306 @@
+        &self.struct_name
+    }
+
+    fn docs(&self) -> &str {
+        &self.docs
+    }
+
+    fn title(&self) -> &str {
+        "Struct"
+    }
+}
+
+struct EnumData {
+    enum_name: String,
+    docs: String,
+}
+
+impl EnumData {
+    fn from_item(item: &syn::ItemEnum) -> Option<Self> {
+        let item = match item.vis {
+            syn::Visibility::Public(_) => item,
+            syn::Visibility::Restricted(_) => return None,
+            syn::Visibility::Inherited => return None,
+        };
+
+        let name = item.ident.to_string();
+
+        let docs = pull_docs_from_attribute(&item.attrs);
+
+        Some(Self {
+            enum_name: name,
+            docs,
+        })
+    }
+
+    fn pull_from_items(all_items: &[syn::Item]) -> Vec<EnumData> {
+        all_items
+            .iter()
+            .filter_map(|item| match item {
+                syn::Item::Enum(item) => EnumData::from_item(item),
+                _ => None,
+            })
+            .collect()
+    }
+}
+
+impl Documentable for EnumData {
+    fn name(&self) -> &str {
+        &self.enum_name
+    }
+
+    fn docs(&self) -> &str {
+        &self.docs
+    }
+
+    fn title(&self) -> &str {
+        "Enum"
+    }
+}
+
+#[derive(Debug)]
+struct FunctionData {
+    function_name: String,
+    docs: String,
+}
+
+impl FunctionData {
+    fn from_item(item: &syn::ItemFn) -> Option<Self> {
+        let item = match item.vis {
+            syn::Visibility::Public(_) => item,
+            syn::Visibility::Restricted(_) => return None,
+            syn::Visibility::Inherited => return None,
+        };
+
+        let name = item.sig.ident.to_string();
+
+        let docs = pull_docs_from_attribute(&item.attrs);
+
+        Some(FunctionData {
+            function_name: name,
+            docs,
+        })
+    }
+
+    fn pull_from_items(all_items: &[syn::Item]) -> Vec<FunctionData> {
+        all_items
+            .iter()
+            .filter_map(|item| match item {
+                syn::Item::Fn(item) => FunctionData::from_item(item),
+                _ => None,
+            })
+            .collect()
+    }
+}
+
+impl Documentable for FunctionData {
+    fn name(&self) -> &str {
+        &self.function_name
+    }
+
+    fn docs(&self) -> &str {
+        &self.docs
+    }
+
+    fn title(&self) -> &str {
+        "Function"
+    }
+}
+
+fn pull_docs_from_attribute(attrs: &[syn::Attribute]) -> String {
+    let docs = attrs
+        .iter()
+        .filter_map(|m| m.meta.require_name_value().ok())
+        .filter(|m| {
+            m.path.segments.first().is_some()
+                && m.path.segments.first().expect("Was checked").ident == "doc"
+        })
+        .filter_map(|v| match &v.value {
+            syn::Expr::Lit(lit) => Some(lit),
+            _ => None,
+        })
+        .filter_map(|l| match &l.lit {
+            syn::Lit::Str(s) => Some(s),
+            _ => None,
+        })
+        .map(|s| s.value().trim().to_string())
```

Comment:
All of this could be just one `filter_map`. Not sure how much improvement would it be though.

Reply:
Something like (just typing here, not guaranteed to work):
```rust
.filter_map(|m| {
    let m = m.meta.require_name_value().ok()?;
    m.path().is_ident("doc").then_some(())?;  // poor man's `ensure!`
    let val = &m.value;
    let str_lit = quote!(#val).parse2::<syn::LitStr>().ok()?;
    Some(str_lit.value().trim().to_string())
})
```